### PR TITLE
design: Application-Level Policies and Customized Control-Logic Workflow

### PR DESCRIPTION
**Pull request by hongchaodeng**
ref https://github.com/oam-dev/kubevela/issues/1613

### CODE REVIEW

**Comment by hongchaodeng** [comment]
@wonderflow Updated the proposal to reflect our discussion today.

File: design/vela-core/workflow_policy.md [hunk 1]
```diff
@@ -0,0 +1,325 @@
+# Application-Level Policies and Customized Control-Logic Workflow Design
+
+## Background
+
+The current model consists of mainly Components and Traits. While this enables the Application object to plug-in operational capabilities, it is still no flexible enough. Specifically, it has the following limitations:
+
+- The current control logic could not be customized. Once the Vela controller renders final k8s resources, it simply applies them without any extension points. In some scenarios, users want to do more complex operations like:
+  - Blue-green upgrade old-new app revisions.
```
[the remark this answers is not in the document]
Comment (hongchaodeng) [reply design/vela-core/workflow_policy.md:8]: This design is based on our existing efforts and experience on many features including rollout trait, appDeploy. This is a continuation of the work, including the rollout trait you mentioned.

The problem with previous rollout and some other features is that they are tightly coupled to the control logic. The workflow design will decouple them and enable customized control logic to plug in. A rollout feature like the previous one will be modeled as a workflowStep.

File: design/vela-core/workflow_policy.md [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+  - User interaction like manual approval/rollback.
+  - Distributing workloads across multiple clusters.
```
[the remark this answers is not in the document]
Comment (hongchaodeng) [reply design/vela-core/workflow_policy.md:10]: > if we're removing the multi-cluster step from the pipeline, is vela supposed to garbage-collect the orphaned resources distributed in the clusters (if so, how is the underlying per-step controller informed of that knowledge) or simply discard them?

Vela will try to remove the WorkflowStep object. The custom controller (e.g. multi-cluster controller) will be responsible for garbage-collecting distributed resources across clusters.


>  if we're flattening them as separate elements in one array, am afraid that bring extra verbosity to the application model as the cost of human-readability

There is no restriction that prevents you from putting them into one item/object.

File: design/vela-core/workflow_policy.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+The process goes as:
+
+- On creating the application, app controller will apply the HelmTemplate/KustomizePatch objects, and wait for its status.
+- The HelmTemplate/KustomizePatch controller would read the template from specified source, render the final config. It will compare the config with the Application object -- if there is difference, it will write back to the Application object per se.
+- The update of Application will trigger another event, the app controller will apply the HelmTemplate/KustomizePatch objects with new context. But this time, the HelmTemplate/KustomizePatch controller will find no diff after the rendering. So it will skip this time.
+
+## Considerations
+
+### Comparison with Argo Workflow/Tekton
```
[the remark this answers is not in the document]
Comment (hongchaodeng) [reply design/vela-core/workflow_policy.md:321]: Yes we are trying to not reinvent the wheel as much as we could. Please see the justification in the section.

File: design/vela-core/workflow_policy.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+Each workflow step has the following interactions with the app controller:
+- The controller will apply the workflow object with annotation `app.oam.dev/workflow-context`. This annotation will pass in the context marshalled in json defined as the following:
+  ```go
+  type WorkflowContext struct {
+    AppName string
+    AppRevisionName string
+    WorkflowIndex int
+  }
+  ```
+- The controller will wait for the workflow object's annotation `app.oam.dev/workflow-phase` to have value:
```
[the remark this answers is not in the document]
Comment (hongchaodeng) [reply design/vela-core/workflow_policy.md]: Fixed.

**Review by wonderflow — APPROVED** [review]

**Pull request merged**